Unify virtual machine documented examples (#7267)
istio/istio.io: content/en/docs/examples/virtual-machines/bookinfo/index.md

---

> title: Examples of using virtual machine integration with Istio

Reviewer: My understanding was that these tutorials were aimed at adding non-k8s services to a mesh. Maybe we should title it as such?

Reviewer: Adding external services to the mesh, perhaps?

As an Istio user, I am much more likely use that as a search query and stumble upon these docs.

> description: Send and receive messages, use Bookinfo application with a MySQL service running on a virtual

Reviewer: ```suggestion
description: Send and receive messages, use the Bookinfo application with a MySQL service running on a virtual
```

- /docs/examples/vm-bookinfo
---

## Send requests from VM workloads to Kubernetes services

After setup, the machine can access services running in the Kubernetes cluster
or on other VMs.

The following example shows accessing a service running in the Kubernetes cluster from a VM using
`/etc/hosts/`, in this case using a service from the [Bookinfo example](/docs/examples/bookinfo/).

> 1.  First, on the cluster admin machine get the virtual IP address (`clusterIP`) for the service:

Reviewer: First is a tad redundant given we are using a numbered list here.

    {{< text bash >}}
    $ kubectl get svc productpage -o jsonpath='{.spec.clusterIP}'
    10.55.246.247
    {{< /text >}}

1.  Then on the added VM, add the service name and address to its `etc/hosts`
    file. You can then connect to the cluster service from the VM, as in the
    example below:

    {{< text bash >}}
$ echo "10.55.246.247 productpage.default.svc.cluster.local" | sudo tee -a /etc/hosts
$ curl -v productpage.default.svc.cluster.local:9080
< HTTP/1.1 200 OK
< content-type: text/html; charset=utf-8
< content-length: 1836
< server: envoy
... html content ...
    {{< /text >}}

> The `server: envoy` header indicates that the sidecar intercepted the traffic.

Reviewer: ```suggestion
The `server: envoy` header indicates that the sidecar has intercepted the traffic.
```

> The `server: envoy` header indicates that the sidecar intercepted the traffic.

Reviewer: ```suggestion
The `server: envoy` header indicates that the sidecar has intercepted the traffic.
```

> ## Running services on the added VM

Reviewer: Maybe something "starting up and running your external service"?

> 1. Setup an HTTP server on the VM instance to serve HTTP traffic on port 8080:

Reviewer: Add a warning note that a GCE account is required here ?

> 1. Setup an HTTP server on the VM instance to serve HTTP traffic on port 8080:

Reviewer 2: ```suggestion
1. Set up an HTTP server on the VM instance to serve HTTP traffic on port 8080:
```
(https://developers.google.com/style/word-list#setup)

    {{< text bash >}}
    $ gcloud compute ssh ${GCE_NAME}
    $ python -m SimpleHTTPServer 8080
    {{< /text >}}

> 1. Determine the VM instance's IP address. For example, find the IP address
>     of the GCE instance with the following commands:

Reviewer 2: ```suggestion
1. Determine the VM instance's IP address. For example, to find the IP address
    of the GCE instance, use the following commands:
```
(Fine to leave it as you have it, but I think it's a little clearer with the "to".)

    {{< text bash >}}
    $ export GCE_IP=$(gcloud --format="value(networkInterfaces[0].networkIP)" compute instances describe ${GCE_NAME})
    $ echo ${GCE_IP}
    {{< /text >}}

> 1. Add VM services to the mesh

Reviewer 2: ```suggestion
1. Add VM services to the mesh:
```

    {{< text bash >}}
    $ istioctl experimental add-to-mesh external-service vmhttp ${VM_IP} http:8080 -n ${SERVICE_NAMESPACE}
    {{< /text >}}

    {{< tip >}}

>     Ensure you have added the `istioctl` client to your path, as described in the [download page](/docs/setup/getting-started/#download).

Reviewer 2: ```suggestion
    Ensure that you have added the `istioctl` client to your path, as described in the [download page](/docs/setup/getting-started/#download).
```

(https://developers.google.com/style/pronouns#optional-pronouns)

    {{< /tip >}}

1. Deploy a pod running the `sleep` service in the Kubernetes cluster, and wait until it is ready:

    {{< text bash >}}
    $ kubectl apply -f @samples/sleep/sleep.yaml@
    $ kubectl get pod
    NAME                             READY     STATUS    RESTARTS   AGE
    sleep-88ddbcfdd-rm42k            2/2       Running   0          1s
    ...
    {{< /text >}}

1. Send a request from the `sleep` service on the pod to the VM's HTTP service:

    {{< text bash >}}
    $ kubectl exec -it sleep-88ddbcfdd-rm42k -c sleep -- curl vmhttp.${SERVICE_NAMESPACE}.svc.cluster.local:8080
    {{< /text >}}

>     You should see something similar to the output below.

Reviewer 2: We recommend avoiding using the word "should" where possible. (https://developers.google.com/style/word-list#should)

    {{< text html >}}
    <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 3.2 Final//EN"><html>
    <title>Directory listing for /</title>
    <body>
    <h2>Directory listing for /</h2>
    <hr>
    <ul>
    <li><a href=".bashrc">.bashrc</a></li>
    <li><a href=".ssh/">.ssh/</a></li>
    ...
    </body>
    {{< /text >}}

> **Congratulations!** You successfully configured a service running in a pod within the cluster to

Reviewer: I prefer the perfect tense here.
```suggestion
**Congratulations!** You have successfully configured a service running in a pod within the cluster to
```

send traffic to a service running on a VM outside of the cluster and tested that
the configuration worked.

## Cleanup

> Run the following commands to remove the expansion VM from the mesh's abstract

Reviewer: Not a big fan of "expansion VM" as an outside reader. I still prefer "external service". We could also call the VM that contains the external service, the "expansion VM" earlier on to fill the reader in maybe?

Author: I see where the challenge is after reading your review with the word 'integrate with Virtual Machine". How about "Integrate with an external service running in a Virtual machine"

model.

{{< text bash >}}
$ istioctl experimental remove-from-mesh -n ${SERVICE_NAMESPACE} vmhttp
Kubernetes Service "vmhttp.vm" has been deleted for external service "vmhttp"
Service Entry "mesh-expansion-vmhttp" has been deleted for external service "vmhttp"
{{< /text >}}

> ## Troubleshooting

Reviewer 2: Should this troubleshooting  section come after the existing material about Bookinfo?

The following are some basic troubleshooting steps for common VM-related issues.

> -    When making requests from a VM to the cluster, ensure you don't run the requests as `root` or
>     `istio-proxy` user. By default, Istio excludes both users from interception.

Reviewer 2: ```suggestion
-    When making requests from a VM to the cluster, ensure that you don't run the requests as the users `root` or
    `istio-proxy`. By default, Istio excludes both users from interception.
```
You don't have to make exactly this change, but I feel like the phrase "`root` or `istio-proxy` user" is a little awkward/confusing, so I recommend changing it a little, even if you don't make exactly the change I'm suggesting.

> -    Verify the machine can reach the IP of the all workloads running in the cluster. For example:

Reviewer 2: ```suggestion
-    Verify that the machine can reach the IP of all of the workloads running in the cluster. For example:
```
Also, is there more than one IP? If so, then change "IP" to "IPs".

    {{< text bash >}}
    $ kubectl get endpoints productpage -o jsonpath='{.subsets[0].addresses[0].ip}'
    10.52.39.13
    {{< /text >}}

    {{< text bash >}}
    $ curl 10.52.39.13:9080
    html output
    {{< /text >}}

-    Check the status of the Istio Agent and sidecar:

    {{< text bash >}}
    $ sudo systemctl status istio
    {{< /text >}}

-    Check that the processes are running. The following is an example of the processes you should see on the VM if you run
     `ps`, filtered for `istio`:

    {{< text bash >}}
    $ ps aux | grep istio
    root      6955  0.0  0.0  49344  3048 ?        Ss   21:32   0:00 su -s /bin/bash -c INSTANCE_IP=10.150.0.5 POD_NAME=demo-vm-1 POD_NAMESPACE=vm exec /usr/local/bin/pilot-agent proxy > /var/log/istio/istio.log istio-proxy
    istio-p+  7016  0.0  0.1 215172 12096 ?        Ssl  21:32   0:00 /usr/local/bin/pilot-agent proxy
    istio-p+  7094  4.0  0.3  69540 24800 ?        Sl   21:32   0:37 /usr/local/bin/envoy -c /etc/istio/proxy/envoy-rev1.json --restart-epoch 1 --drain-time-s 2 --parent-shutdown-time-s 3 --service-cluster istio-proxy --service-node sidecar~10.150.0.5~demo-vm-1.vm-vm.svc.cluster.local
    {{< /text >}}

> -    Check the Envoy access and error logs:

Reviewer: We could link the reader to the troubleshooting wiki here
